source/common/http/http1/codec_impl.cc

```diff
@@ -247,6 +247,11 @@ void RequestStreamEncoderImpl::encodeHeaders(const HeaderMap& headers, bool end_
 
   if (method->value() == Headers::get().MethodValues.Head.c_str()) {
     head_request_ = true;
+    // If this is an incomplete request (end_stream=false), then connection_.onEncodeComplete();
+    // will not be called in StreamEncoderImpl::encodeHeaders, head request state will not be passed
+    // to the corresponding response so here need to call this method
+    if (!end_stream)
```

Comment:
nit: needs braces around this if statement. 

I'm still not a fan of calling `onEncodeComplete` here as the encode is not necessarily complete. Can we just pass the state directly? This is what I was referring to previously. Can the the RequestStreamEncoderImpl hold a reference to the ClientConnection such that we can transfer this state directly? I think that might be cleaner?

Reply:
You actually think that the name of onEncodeComplete will have some ambiguity. In fact, there is no complete code completion. 

Can I directly dynamic_cast from ConnectionImpl to ClientConnectionImpl and pass the state directly?